src/Microsoft.AspNetCore.Sockets.Client/WebSocketsTransport.cs

```diff
@@ -0,0 +1,153 @@
+        {
+            _logger = loggerFactory.CreateLogger<WebSocketsTransport>();
+        }
+        public Task Running { get; private set; }
+                                                
+        public async Task StartAsync(Uri url, IChannelConnection<Message> application)  
+        {
+            if (url == null)
+            {
+                throw new ArgumentNullException(nameof(url));
+            }
+
+            if (application == null)
+            {
+                throw new ArgumentNullException(nameof(application));
+            }
+
+            _application = application;
+            await Connect(url);
+            var sendTask = SendMessages(url, _cancellationToken);
+            var receiveTask = ReceiveMessages(url, _cancellationToken);
+
+            Running = Task.WhenAll(sendTask, receiveTask).ContinueWith(t => {
+                _application.Output.TryComplete(t.IsFaulted ? t.Exception.InnerException : null);
+                return t;
+            }).Unwrap();
+        }
+
+        private async Task ReceiveMessages(Uri pollUrl, CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                const int bufferSize = 1024;
+                var totalBytes = 0;
+                var incomingMessage = new List<ArraySegment<byte>>();
+                WebSocketReceiveResult receiveResult;
+                do
+                {
+                    var buffer = new ArraySegment<byte>(new byte[bufferSize]);
+                    receiveResult = await _webSocket.ReceiveAsync(buffer, cancellationToken);
+                    var truncBuffer = new ArraySegment<byte>(new byte[receiveResult.Count]);
+                    Buffer.BlockCopy(buffer.Array, 0, truncBuffer.Array, 0, receiveResult.Count);
+
+                    incomingMessage.Add(truncBuffer);
+                    totalBytes += receiveResult.Count;
+                } while (!receiveResult.EndOfMessage);
+
+                Message message;
+                if (incomingMessage.Count > 1)
+                {
+                    var messageBuffer = new byte[totalBytes];
+                    var offset = 0;
+                    for (var i = 0 ; i < incomingMessage.Count; i++)
+                    {
+                        Buffer.BlockCopy(incomingMessage[i].Array, 0, messageBuffer, offset, incomingMessage[i].Count);
+                        offset += incomingMessage[i].Count;
+                    }
+                    
+                    message = new Message(ReadableBuffer.Create(messageBuffer).Preserve(), receiveResult.MessageType == 0 ? Format.Text : Format.Binary, receiveResult.EndOfMessage);
+                }
+                else
+                {
+                    message = new Message(ReadableBuffer.Create(incomingMessage[0].Array).Preserve(), receiveResult.MessageType == 0 ? Format.Text : Format.Binary, receiveResult.EndOfMessage);
```

Comment:
Use the overload of `Create` which takes an offset and a length and use the `Offset` and `Count` properties from `incomingMessage[0]`. `ArraySegment.Array` gives you the **entire** array, not just the specific slice that ended up having data, so we need to make sure to pass the offset and count along everywhere we use it.